### DISCUSSION TITLE
Allotment.Pane size

### DISCUSSION THREAD

**Discussion by smu-cns** [johnwalley/allotment#57] [Q&A]
Hello John, 

I am creating an editor and want to add a new Allotment.Pane dynamically from an array. However, doing so will render the new pane with the minimum size only (30px). When I double click on the sash, the pane is resized correctly. Is there any way to force this? I looked into the code and it seems adding a new pane should distribute the panes correctly. Any idea?

Thanks,

Stefan

**Comment by johnwalley** [comment]
Hi Stefan. This is a perfect use case for Allotment (it's based on VS Code after all) so it definitely **should** work. However, as you've discovered, it doesn't.

Evenly distributing the space between the panes feels like a good default. It's how the VS Code editor panes behave.

![Nov-25-2021 10-16-09](https://user-images.githubusercontent.com/981531/143422992-52a480fb-0575-454d-8cf9-58527b9b9ca0.gif)

And as you point out, the code does look like it should do this.

<img width="365" alt="Screenshot 2021-11-25 at 10 17 51" src="https://user-images.githubusercontent.com/981531/143423208-b5d9b52a-7c7c-4b9c-88f4-01d1dede6651.png">

I'll aim to fix this next opportunity I have, probably tomorrow. I'll let you know when I publish an updated version.

**Reply by smu-cns** [reply]
That would be really great. Thank you John!

**Reply by johnwalley** [reply]
I've published [v1.2.1](https://www.npmjs.com/package/allotment/v/1.2.1) with a fix. If you could give it a try and let me know how you get on that would be great.

Thanks

**Reply by smu-cns** [reply]
Yes - adding works as expected. Thank you! 

There is an issue now when removing the pane. The space will not get reassigned. The middle pane was removed in the screenshot:

![image](https://user-images.githubusercontent.com/80399264/143454856-ddac8763-13df-44f0-a0ac-9fd007193c62.png)

When you remove the view, the space should be redistributed however. In removeView, the last lines are:

```
if (sizing && sizing.type === "distribute") {
      this.distributeViewSizes();
}
```

**Reply by johnwalley** [reply]
I'll have a look at this too. I'm missing test cases for removing panes in the middle so first thing I'll do i add them and see what needs to be done.

**Reply by smu-cns** [reply]
Thank you again. In 1.2.0 this did not happen reproducible. So somehow the internal state seems to be messed up from the additional layouting.

**Comment by johnwalley** [comment]
Do you have an example I could look at? I'm struggling to reproduce it in the hastily put together test I've created.

![Nov-25-2021 14-44-11](https://user-images.githubusercontent.com/981531/143461528-c2312544-c39b-403a-ad0a-b1527a4ab757.gif)

I'm offline for the rest of the day but will take another look tomorrow.

**Reply by smu-cns** [reply]
I will try to create one. My steps to reproduce are:

- Create a single allotment pane 
- Add 3 tabs within the same pane
- "Split" the first tab (make a new allotment pane)
- "Split" the second tab (make a new allotment pane)
- Close the second or third pane.

So maybe the adding and removing afterwards messes up the state? 
As you can see in my screenshot, I am using a material tab within the pane. 
Maybe I have an error in my code - have to check this.

**Reply by smu-cns** [reply]
Hi John,

I made a stackblitz: https://react-omir7g.stackblitz.io
To edit: https://stackblitz.com/edit/react-omir7g?file=demo.tsx

It is not exactly what I do but it uses a layout around the allotment. Click a few times "Add". Then close the panels starting at the second one. I would expect the panels to reoccupy the remaining space (they do when you double click on the sash).

This is not the same behavior I am seeing in my app (as you can see in the screenshot above). 
But I guess it is related?

**Reply by johnwalley** [reply]
Perhaps we've been talking about different things! Does the following describe what you're seeing?

1. You remove an allotment pane
2. The remaining panes take up 100% of the container width but are not equally distributed? In particular, the last pane grows to take up the missing space.

If so then that's the default behaviour but I'm happy to change it to equally distribute the pane sizes when one (or more) is removed. It's not obvious what the behaviour should be when adding and removing panes but resetting the layout and distributing them as evenly as possible (modulo min/max sizes) sounds like a good default to me.

If instead you're seeing the panes not add up to 100% of the container width then that's a bug but I can't reproduce.

And thanks for taking the time to put together the stackblitz.

**Reply by smu-cns** [reply]
I think I have to adapt the sample again. If I think about it, even in my example the positioning looks correct..

**Reply by smu-cns** [reply]
Ah sorry I did not see your answer. I will do another stackblitz - just give me some time (hopefully Monday).

**Reply by smu-cns** [reply]
The two steps you mention describe what I am seeing in my stackblitz example. I thought this would be the bug I observed in my app - but this not the case. The panes widths add up to 100%, as you mentioned. If I had colored the panes differently I would have seen it.

So my problem is what you mention last: 
In my screenshot the chat pane was closed. Neither of the other panels grabbed the space as expected. The panel widths do not add up to 100%.

I am struggling to reproduce the problem (in an stackblitz example) as well. My suspicion was, the layout of my application around the Allotment could have an influence because in my app it is reproducible 100% of the time when there are >= 3 panes. With 2 panes it is working fine.

However, no luck yet..

**Reply by smu-cns** [reply]
Hello John, 

We are currently restructuring our App. Maybe the issue resolves or we come closer to what the actual problem is so I can provide a new Stackblitz.

**Reply by johnwalley** [reply]
Thanks for keeping me up to date. I'll keep on eye out for the problem in the meantime with my own testing.

**Reply by smu-cns** [reply]
I could reproduce the issue. The problem is, that I do not add the panels in linear order as it is done in my previous demo and in your test cases. This will mess up the order of React-Components. You can look here:

https://945se.csb.app/

When you press the button in the action bar in the left, a new allotment is added to the right of the rightmost allotment. Opening and closing works fine.

However, in my app I use something I call "Split". When you press the button within a panel, the new allotment should be shown right to the allotment being clicked not next to the rightmost. 

In the following screenshot, I created the panels `Panel-T0w` and `Panel-Xdg` using the button in the action bar. Then I used the split button of `Panel-T0w`. Allotment `Panel-Wv` was added and is rendered as second. However, in the component tree it is the last one. So I guess the resizing logic will assume some fixed order because closing this `Panel-Wv` will lead to the error.

![image](https://user-images.githubusercontent.com/80399264/149373893-b12ad448-d12a-409c-a03a-fc759c23a880.png)

I am not sure if this is now my problem or in the lib. But at least I can reproduce it.

PS: The example setup was rather complex because I tried to mimic what I have here in order to make sure it is not some styling problem.


**EDIT:**
We had a longer debug session and just figured, that split would work if we do not assign a key to the Allotment.Pane being created. In other words, if we reassign the keys after split it works. The drawback is, that this will cause all child components of a pane to be recreated. I will see how/if we can work around this.

**Reply by johnwalley** [reply]
Thanks for taking the time to put this together. I'll start investigating it today.

**Reply by johnwalley** [reply]
I've published a new version with a fix: https://www.npmjs.com/package/allotment/v/1.5.2.

I was failing to maintain the ordering between the React components and the internal view items. Bit of a blind spot made worse by all my test cases never changing the order!

If you could let me know if this fixes your issue when you get the chance that'd be great 🙏 .

**Reply by smu-cns** [reply]
Thanks for the fast response and the release.
Unfortunately the issue is still there.  I tried in my real application and in the sandbox.

**Reply by johnwalley** [reply]
Could you try `1.6.0`? I think I included the wrong build in `1.5.2` 😕

**Reply by smu-cns** [reply]
I tried in my sandbox - Looks good! 
Will try it in the app later this day.

Thank you!

**Reply by smu-cns** [reply]
Hello John - it works like charm!
